File: src/arabagent/WikiData.java
package arabagent;

import java.util.*;
import java.io.*;
import java.net.URL;


public class WikiData {
	//static String files_path = "C:/Documents and Settings/disooqi/workspace/ArabAgent/war/wikiData/";

    /*
     * 0 = unloaded
     * 1 = loading
     * 2 = loaded
     * */
    static int is_loaded = 0;
    
    //classLoader.
    //ClassLoader().;//.getResource("_cc_inconcepts_count.txt");
    //static URL inlink_counts_path = ClassLoader().getResource("_cc_inconcepts_count.txt");
    //static URL term_concepts_path = WikiData.class.getClassLoader().getResource("term_concepts_list_index.txt");

    static String inlink_counts_path = "_cc_inconcepts_count.txt";
    static String concept_outconcepts_list_index = "_concept_outconcepts_list_index.txt";
    static String concept_inconcepts_list_index = "_concept_inconcepts_list_index.txt";
    static String term_concepts_path = "term_concepts_list_index.txt";
    
    static Map<String, String> concept_outconcepts_index = new HashMap<String, String>();
    static Map<String, String> concept_inconcepts_index = new HashMap<String, String>();
    static Map<String, Double> inlink_counts_cache = new HashMap<String, Double>();
    static Map<String, String> term_concepts_dic = new HashMap<String, String>();
    
    public static void load_wiki_files()
    {
        switch (is_loaded)
        {
            case 0:
                is_loaded = 1;
                break;
            case 1:
                while (is_loaded == 1) ;
                return;
            case 2: return;
        }
        
        try
        {
        	InputStream is = WikiData.class.getResourceAsStream(inlink_counts_path);
        	//InputStream is = WikiData.class.getClassLoader().getResourceAsStream(inlink_counts_path);        	        	
        	BufferedReader br = new BufferedReader(new InputStreamReader(is, "UTF8"));

            String word = "";
            while ((word = br.readLine()) != null)            
              inlink_counts_cache.put(word.substring(0, word.indexOf(',')), Double.parseDouble(word.substring(word.indexOf(',') + 1)));
            is.close();
            br.close();
        	/*  
            //to create the feature vector of the category (41M)
            fis = new FileInputStream(concept_outconcepts_list_index);        	        	
        	br = new BufferedReader(new InputStreamReader(fis, "UTF8"));

            word = "";
            while ((word = br.readLine()) != null)                
                concept_outconcepts_index.put(word.substring(0, word.indexOf(':')), word.substring(word.indexOf(':') + 1));
            fis.close();  
            br.close();
                            
            //for google similarity distance (33M)
            fis = new FileInputStream(concept_inconcepts_list_index);        	        	
            br = new BufferedReader(new InputStreamReader(fis, "UTF8"));

            word = "";          
            while ((word = br.readLine()) != null)            
                concept_inconcepts_index.put(word.substring(0, word.indexOf(':')), word.substring(word.indexOf(':') + 1));
            fis.close(); 
            br.close();*/ 
           
            //15M
            is = WikiData.class.getResourceAsStream(term_concepts_path);        	        	
            br = new BufferedReader(new InputStreamReader(is, "UTF8"));

            word = "";
            while ((word = br.readLine()) != null)     
                term_concepts_dic.put(word.substring(0, word.indexOf(':')), word.substring(word.indexOf(':') + 1));
            is.close(); 
            br.close();
         
            is_loaded = 2;
        }
        catch (FileNotFoundException e) { e.getMessage(); }
        catch (UnsupportedEncodingException e){e.getMessage();}
        catch (IOException e){e.getMessage();}
    }

    public static String get_concept_wikilinks(String concept_id)
    {
        return concept_outconcepts_index.get(concept_id);
    }

    public static String what_links_here(String concept_id)
    {
        return concept_inconcepts_index.get(concept_id);
    }
    
    public static String get_concepts_for_a_term(String term)
    {        
        if(term_concepts_dic.containsKey(term))
        	return term_concepts_dic.get(term);
		return "";
    }

    public static double get_inlinks_count(String c_id)
    {
        return inlink_counts_cache.get(c_id);
    }


}

File: DataAccess/WikiData.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Xml;

namespace DataAccess
{
    public static partial class WikiData
    {
        static string files_path = @"D:\Dropbox\_workspace\work\KSA research group\ArabAgent\data\";
        static string is_loaded = "unloaded";

        //static string inlink_counts_path = "_cc_inlink_count(sorted).sql";
        //static string concept_outconcepts_list_index = "_concept_outconcepts_list_index(sorted).sql";
        //static string concept_inconcepts_list_index = "_concept_inconcepts_list_index(sorted).sql";
        //static string term_concepts_path = "_term_concepts(norm_no_stopwords).sql";

        static string inlink_counts_path = "_cc_inconcepts_count.sql";
        static string concept_outconcepts_list_index = "_concept_outconcepts_list_index.sql";
        static string concept_inconcepts_list_index = "_concept_inconcepts_list_index.sql";
        static string term_concepts_path = "term_concepts_list_index.sql";


        static Dictionary<string, string> concept_outconcepts_index = new Dictionary<string, string>();
        static Dictionary<string, string> concept_inconcepts_index = new Dictionary<string, string>();
        static Dictionary<string, double> inlink_counts_cache = new Dictionary<string, double>();
        static Dictionary<string, string> term_concepts_dic = new Dictionary<string, string>();

        //public static string Files_path
        //{
        //    set { files_path = value; }
        //    get { return files_path; }
        //}

        static WikiData() {  }

        public static void load_wiki_files()
        {
            switch (is_loaded)
            {
                case "unloaded":
                    is_loaded = "loading";
                    break;
                case "loading":
                    while (is_loaded == "loading") ;
                    return;
                case "loaded": return;
            }

            try
            {
                using (StreamReader sr = new StreamReader(files_path+inlink_counts_path, Encoding.UTF8))
                {
                    string word = string.Empty;
                    while ((word = sr.ReadLine()) != null)
                    {
                        inlink_counts_cache.Add(word.Substring(0, word.IndexOf(',')), double.Parse(word.Substring(word.IndexOf(',') + 1)));
                    }
                }
                //to create the veture vector of the category (41M)
                using (StreamReader sr = new StreamReader(files_path + concept_outconcepts_list_index, Encoding.UTF8))
                {
                    string word = string.Empty;
                    while ((word = sr.ReadLine()) != null)
                    {
                        concept_outconcepts_index.Add(word.Substring(0, word.IndexOf(':')), word.Substring(word.IndexOf(':') + 1));
                    }
                }
                //for google similarity distance (33M)
                using (StreamReader sr = new StreamReader(files_path + concept_inconcepts_list_index, Encoding.UTF8))
                {
                    string word = string.Empty;
                    while ((word = sr.ReadLine()) != null)
                    {
                        concept_inconcepts_index.Add(word.Substring(0, word.IndexOf(':')), word.Substring(word.IndexOf(':') + 1));
                    }
                }

                //15M
                using (StreamReader sr = new StreamReader(files_path + term_concepts_path, Encoding.UTF8))
                {
                    string word = string.Empty;
                    while ((word = sr.ReadLine()) != null)
                    {
                        term_concepts_dic.Add(word.Substring(0, word.IndexOf(':')), word.Substring(word.IndexOf(':') + 1));
                    }
                }

                is_loaded = "loaded";
            }
            catch (Exception load_exp)
            {
                throw new Exception("Ya Disooqi: " + load_exp.Message, load_exp);
            }
        }
        
        public static string get_concept_wikilinks(string concept_id)
        {
            return concept_outconcepts_index[concept_id];
        }

        public static string what_links_here(string concept_id)
        {
            return concept_inconcepts_index[concept_id];
        }
        
        public static string get_concepts_for_a_term(string term)
        {
            string temp = string.Empty;
            term_concepts_dic.TryGetValue(term, out temp);
            return temp;
        }

        public static double get_inlinks_count(string c_id)
        {
            return inlink_counts_cache[c_id];
        }

        //public static int number_of_article_that_mentioned_the_phrase_at_all(string c_id)
        //{
        //    return context_term_appearance[c_id];
        //}

        //public static void reload_term_concepts_index()
        //{
        //    using (StreamReader sr = new StreamReader(term_concepts_path, Encoding.UTF8))
        //    {
        //        string word = string.Empty;
        //        while ((word = sr.ReadLine()) != null)
        //        {
        //            term_concepts_dic.Add(word.Substring(0, word.IndexOf(':')), word.Substring(word.IndexOf(':') + 1));
        //        }
        //    }
        //}
    }
}
